### PR TITLE
wp-scripts: add support for Playwright

### PR DESCRIPTION
**Pull request by swissspidy**
## What?

Adds `wp-scripts test-playwright` with a default Playwright configuration for running e2e tests.

Adds a new `test-playwright` config to `@wordpress/eslint-plugin`.

> [!NOTE]
> Idea: enhance `test-e2e` to automatically detect whether the given configuration is for Puppeteer or Playwright. Needs to accommodate for setups using both together.

## Why?

[Playwright is the future for e2e tests in WordPress](https://make.wordpress.org/core/2022/03/23/migrating-wordpress-e2e-tests-to-playwright/). Adding support directly to wp-scripts makes it easier to use Playwright not just for cure, but for everyone.

## How?

The setup really follows the one for `test-e2e`, though things are a bit simpler because no Jest is involved.

## Testing Instructions

`npm run test:e2e:playwright` and `test:performance:playwright` are changed in the PR to use wp-scripts and the tests still run & pass

You can also run `./node_modules/.bin/wp-scripts test-playwright` and then provide a different config to run tests in another project.

### CODE REVIEW

**Comment by gziolo** [comment]
Thank you for starting this PR and exploring how we can add support for Playwright to `@wordpress/scripts`. It will eventually fully replace Puppeteer, the default tool behind `wp-scripts test-e2e`.

> Idea: enhance test-e2e to automatically detect whether the given configuration is for Puppeteer or Playwright. Needs to accommodate for setups using both together.

It sounds like a great idea. If folks are fine using the default config provided by `@wordpress/scripts` then it should default to Playwright for new projects and probably continue using Puppeteer for existing projects. Not sure how difficult technically that would be, though, or even if that is necessary.

---
> Blockers
> Need to install browsers

I see that we install browsers on CI with

```sh
npx playwright install chromium firefox webkit --with-deps
```

You can replicate how we install browsers when running `test-e2e` with Puppeteer:

https://github.com/WordPress/gutenberg/blob/422d12ad49fe28b91d22994c82a5c5dd96f8474f/packages/scripts/scripts/test-e2e.js#L31-L33

This way it magically works for the dev when they run it for the first time.

---

My biggest concern was initially about adding new dependence for Playwright, while we have dependencies present for Puppeteer in `@wordpress/scripts`. I did some quick calculations in the newly created project to check the impact:

Screenshot 2023-08-11 at 11 32 30
Screenshot 2023-08-11 at 11 32 52

Both tools have a similar impact, a little bit over 15 MB which isn't surprising as Playwright was forked from Puppeteer. The great news is that both are optimized and they don't install browsers together with npm packages. It all happens afterward lazily when you actually use the tool.

---

At some point, we will have to revisit the architecture of `@wordpress/scripts` so more stuff is optional and installed only when people use it for the first time. It's particularly important for e2e tooling as I would be surprised to see after the transition period in Gutenberg ends that any project uses both Playwright and Puppeteer at the same time.

As of today, when you install `@wordpress/scripts` you will see 300 MB in your `node_modules` folder. Additional 15 MB doesn't make that bigger difference in this case, but we should aim to decrease the size to a reasonable size so people can install it quicker and save bandwidth when they play with the block development during events like WordPress Contributor Day. The ideal scenario would be to extract Puppeteer to something like `@wordpress/scripts-puppeteer` and ask folks to install it when they want to use it. We could offer something similar for Playwright in case people prefer it as we see it now in WordPress core and Gutenberg now. 

As for this PR, I guess it's better to start with a simpler approach and live for the time being with two concurrent tools bundled with `@wordpress/scripts` at the same time.


Now, test-e2e could do what you suggested - detect a config and run a corresponding tool through these packages. We could even set these extracted packages as optional peer dependencies, and possibly even offer a prompt helping to install them on the fly.

File: packages/eslint-plugin/CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@
 ### Enhancement
 
 -   Support Typescript 5 and 5.1 by updating both `@typescript-eslint/parser` and `@typescript-eslint/eslint-plugin` to version `^5.62.0`. ([#52621](https://github.com/WordPress/gutenberg/pull/52621)).
+-   Added a new `test-playwright` ruleset using [`eslint-plugin-playwright`](https://www.npmjs.com/package/eslint-plugin-playwright).
```
Comment (gziolo) [inline packages/eslint-plugin/CHANGELOG.md]: Nit: it should be moved to the unreleased section.

Aside: we really need a way better way of maintaining changelogs.

File: packages/scripts/CHANGELOG.md [hunk 2]
```diff
@@ -6,9 +6,10 @@
 
 ### Enhancement
 
--   Updated `npm-package-json-lint` peer dependency to require v6.0.0 [#53636](https://github.com/WordPress/gutenberg/pull/53636)
+-   Updated `npm-package-json-lint` peer dependency to require v6.0.0 [#53636](https://github.com/WordPress/gutenberg/pull/53636).
 -   The bundled `@svgr/webpack` dependency has been updated from requiring ^6.2.1 to requiring ^8.0.1 ([#53630](https://github.com/WordPress/gutenberg/pull/53630)).
 -   The bundled `cssnano` dependency has been updated from requiring ^5.07 to requiring ^6.0.1 ([#53630](https://github.com/WordPress/gutenberg/pull/53630)).
+-   Added support for `test-playwright` script ([#53108](https://github.com/WordPress/gutenberg/pull/53108)).
```
Comment (gziolo) [inline packages/scripts/CHANGELOG.md]: Same as above, this entry should be now in unreleased section.

File: packages/scripts/config/playwright/global-setup.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/**
+ * External dependencies
+ */
+import { request } from '@playwright/test';
+import type { FullConfig } from '@playwright/test';
+
+/**
+ * WordPress dependencies
+ */
+import { RequestUtils } from '@wordpress/e2e-test-utils-playwright';
```
Comment (gziolo) [inline packages/scripts/config/playwright/global-setup.ts:10]: I think, it should be defined as a dependency for the package.

**Review by gziolo — COMMENTED** [review]
The changes applied to the ESLint plugin and WP scripts look good.

I don’t know much about Playwright configuration in Gutenberg, but everythings seems to work correctly.

File: packages/eslint-plugin/CHANGELOG.md (hunk 1 above)
Comment (swissspidy) [reply packages/eslint-plugin/CHANGELOG.md]: Good catch! It was there, but after the latest release this got automatically changed without me noticing. Fixed now.

File: .eslintrc.js [hunk 4]
```diff
@@ -308,7 +308,6 @@ module.exports = {
 			rules: {
 				'@wordpress/no-global-active-element': 'off',
 				'@wordpress/no-global-get-selection': 'off',
-				'playwright/no-page-pause': 'error',
```
Comment (gziolo) [inline .eslintrc.js:311]: What's the reason this rule got removed? Is that expected?

File: test/e2e/specs/editor/plugins/nonce.spec.js [hunk 5]
```diff
@@ -16,6 +16,7 @@ test.describe( 'Nonce', () => {
 		await admin.createNewPost();
 		await page.keyboard.press( 'Enter' );
 		// Wait until the network is idle.
+		// eslint-disable-next-line playwright/no-networkidle
```
Comment (gziolo) [inline test/e2e/specs/editor/plugins/nonce.spec.js:19]: Someone could verify that these exceptions caught after the ESLint package upgrade should be refactored in the follow-up PR.

**Review by gziolo — APPROVED** [review]
This looks great as it stands for the `@wordpress/scripts` and `@wordpress/eslint-plugin` with all the remarks I shared in https://github.com/WordPress/gutenberg/pull/53108#issuecomment-1674504210.

@kevin940726 or @WunderBart, could you help verify that refactoring to ESLint rules and configs work as intended?

File: .eslintrc.js (hunk 4 above)
Comment (swissspidy) [reply .eslintrc.js:311]: This rule is already enabled in the recommended ruleset of the Playwright ESLint plugin. No need to specify it separately.

**Comment by swissspidy** [comment]
What's blocking this from reviewing / merging?

**Comment by gziolo** [comment]
npm packages got published to npm today, so I say we waited enough to gather additional feedback. We will have two weeks to adjust if necessary before another publishing window.

**Comment by kevin940726** [comment]
Thanks for the work! FWIW, I didn't want to move Playwright into `wp-scripts` because it's really a standalone opt-in testing tool. I feel like it's way easier for users to decide which testing framework they want to use and follow the tool's documentation. This way, we don't have to maintain yet another documentation for the bundled Playwright dependency and the usages. We can simply guide our users to the official Playwright website if they want to add e2e tests to their projects. This also prevents any potential inconsistencies of config resolution and CLI parsing, for instance. AFAIK, none of the "zero-config" tools/frameworks out there are bundling e2e testing frameworks because they are decoupled from the app. IMHO, copying configs and following the official documentation is way simpler and easier to maintain than rolling our own rules, especially since we have limited resources (and/or priority) to maintain this.

I don't feel that strongly about this or we have to revert this though. Just want to share my opinions for future reference 🙂 .

**Pull request merged**